File: flutter/plugin/lib/core/xplpc.dart
import 'dart:ffi' as ffi;
import 'dart:io';

import 'package:xplpc/core/config.dart';
import 'package:xplpc/proxy/platform_proxy.dart';

class XPLPC {
  static XPLPC? _instance;
  XPLPC._();
  static XPLPC get instance => _instance ??= XPLPC._();

  late ffi.DynamicLibrary library;

  bool initialized = false;
  late Config config;

  void initialize(Config config) {
    if (initialized) {
      return;
    }

    this.config = config;

    library = _openLibrary();
    PlatformProxy.initialize();

    // This is only set once the library is loaded and the proxy is bound, since a failure above must not leave a half built singleton.
    initialized = true;
  }

  ffi.DynamicLibrary _openLibrary() {
    final path = Platform.environment["XPLPC_LIBRARY_PATH"];

    if (path != null && path.isNotEmpty) {
      return ffi.DynamicLibrary.open(path);
    }

    if (Platform.isAndroid || Platform.isLinux) {
      return ffi.DynamicLibrary.open("libxplpc.so");
    }

    if (Platform.isWindows) {
      return ffi.DynamicLibrary.open("xplpc.dll");
    }

    return ffi.DynamicLibrary.process();
  }
}

File: python/lib/src/xplpc/core/xplpc.py
import threading

from xplpc.core.config import Config
from xplpc.proxy.platform_proxy import PlatformProxy


class XPLPC:
    _instance = None
    _instance_lock = threading.Lock()

    def __new__(cls):
        if cls._instance is None:
            with cls._instance_lock:
                if cls._instance is None:
                    cls._instance = super().__new__(cls)
                    cls._instance._init()
        return cls._instance

    def _init(self):
        self.lock = threading.Lock()
        self.initialized = False
        self.config = None

    def initialize(self, config: Config):
        with self.lock:
            if self.initialized:
                return

            self.config = config

            PlatformProxy().initialize()

            # This is only set once the native library is bound, since a failure above must not leave a half built singleton.
            self.initialized = True

    def is_initialized(self):
        return self.initialized
